packages/next/build/webpack/plugins/next-esm-plugin.ts

```diff
@@ -213,38 +213,53 @@ export default class NextEsmPlugin implements Plugin {
       compilation.namedChunkGroups
     )
 
+    const unnamedChunks: compilation.Chunk[] = []
     const childChunkFileMap = childCompilation.chunks.reduce(
       (
         chunkMap: { [key: string]: compilation.Chunk },
         chunk: compilation.Chunk
       ) => {
-        chunkMap[chunk.name] = chunk
+        // Dynamic chunks may not have a name. It'll be null in such cases
+        if (chunk.name === null) {
+          unnamedChunks.push(chunk)
+        } else {
+          chunkMap[chunk.name] = chunk
+        }
+
         return chunkMap
       },
       {}
     )
 
-    // Merge files from similar chunks
+    // Merge chunks - merge the files of chunks with the same name
     compilation.chunks.forEach((chunk: compilation.Chunk) => {
       const childChunk = childChunkFileMap[chunk.name]
 
-      if (childChunk && childChunk.files) {
+      // Do not merge null named chunks since they are different
+      if (chunk.name !== null && childChunk && childChunk.files) {
         delete childChunkFileMap[chunk.name]
         chunk.files.push(
           ...childChunk.files.filter((v: any) => !chunk.files.includes(v))
         )
       }
     })
 
-    // Add modern only chunks
-    compilation.chunks.push(...Object.values(childChunkFileMap))
+    // Add modern only chunks into the main compilation
+    compilation.chunks.push(
+      ...Object.values(childChunkFileMap),
+      ...unnamedChunks
+    )
 
-    // Place modern only chunk inside the right entry point
+    // Place modern only (unmerged) chunks inside the right entry point
     compilation.entrypoints.forEach((entryPoint, entryPointName) => {
       const childEntryPoint = childCompilation.entrypoints.get(entryPointName)
 
       childEntryPoint.chunks.forEach((chunk: compilation.Chunk) => {
-        if (childChunkFileMap.hasOwnProperty(chunk.name)) {
+        if (
+          // Add null named dynamic chunks since they weren't merged
+          chunk.name === null ||
+          childChunkFileMap.hasOwnProperty(chunk.name)
+        ) {
```

Comment:
Question around this if block.

`const childEntryPoint = childCompilation.entrypoints.get(entryPointName)`
If the childEntryPoint has same name as parent compilation's `entryPointName` then why can we just push all `childEntryPoint.chunks` into `entryPoint.chunks` un-conditionally?